okhttp/src/jvmMain/kotlin/okhttp3/internal/connection/ConnectPlan.kt

```diff
@@ -160,6 +160,14 @@ class ConnectPlan(
       }
 
       if (route.address.sslSocketFactory != null) {
+        // Assume the server won't send a TLS ServerHello until we send a TLS ClientHello. If
+        // that happens, then we will have buffered bytes that are needed by the SSLSocket!
+        // This check is imperfect: it doesn't tell us whether a handshake will succeed, just
+        // that it will almost certainly fail because the proxy has sent unexpected data.
+        if (source?.buffer?.exhausted() == false || sink?.buffer?.exhausted() == false) {
+          throw IOException("TLS tunnel buffered too many bytes!")
+        }
```

Comment:
FYI @yschimke I had to move this check to get tests to stop flaking. In TLS handshakes the server doesn’t send until it receives data from the client. In HTTP/2 the server and client can both send immediately!